### ISSUE TITLE
حصه

### ISSUE THREAD

**Issue by yeshesa** [yeshesa/customer_classification.ipynb#3]
"""
bank_shap_explain_fixed.py
نسخة مصححة ومرنة: preprocessing -> RandomForest(GridSearch) -> SHAP (global + local)
سلوك بديل إذا كانت مكتبات الرسم غير مثبتة (matplotlib/seaborn أو plotly).
غيّر INPUT_FILE إلى مسار ملف الإكسل لديك إذا لزم.
"""

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV, StratifiedKFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.metrics import roc_auc_score, f1_score, recall_score, precision_score, accuracy_score
import warnings
import os
import sys
import json
shap (مطلوب)
try:
import shap
except Exception:
print("خطأ: مكتبة shap غير مثبتة. نفّذ: python -m pip install shap")
raise
محاولة استيراد مكتبات الرسم بشكل آمن (fallback إذا لم تكن مثبتة)
HAS_MATPLOTLIB = False
HAS_SEABORN = False
HAS_PLOTLY = False

try:
    import matplotlib.pyplot as plt
    HAS_MATPLOTLIB = True
except Exception:
    HAS_MATPLOTLIB = False

try:
    import seaborn as sns
    HAS_SEABORN = True
except Exception:
    HAS_SEABORN = False

try:
    import plotly.express as px
    import plotly.graph_objects as go
    HAS_PLOTLY = True
except Exception:
    HAS_PLOTLY = False

warnings.simplefilter(action="ignore", category=FutureWarning)
warnings.simplefilter(action="ignore", category=UserWarning)

imblearn اختياري
try:
    from imblearn.over_sampling import SMOTE
except Exception:
    SMOTE = None

shap
try:
    import shap
except Exception:
    print("خطأ: مكتبة shap غير مثبتة. نفّذ: python -m pip install shap")
    raise

محاولة استيراد أدوات الرسم بمرونة
HAS_MATPLOTLIB = False
HAS_SEABORN = False
HAS_PLOTLY = False
plt = None
sns = None
plotly = None

try:
    import matplotlib.pyplot as plt
    HAS_MATPLOTLIB = True
except Exception:
    HAS_MATPLOTLIB = False

try:
    import seaborn as sns
    HAS_SEABORN = True
except Exception:
    HAS_SEABORN = False

try:
    import plotly.express as px
    import plotly.graph_objects as go
    HAS_PLOTLY = True
except Exception:
    HAS_PLOTLY = False

warnings.simplefilter(action="ignore", category=FutureWarning)
warnings.simplefilter(action="ignore", category=UserWarning)

------------------------- متغيرات --------------------------------
INPUT_FILE = "bank_customers.xlsx"   غيّر المسار هنا
SHEET_NAME = 0
RANDOM_STATE = 42
TEST_SIZE = 0.2

TARGET_CANDIDATES = ["Churn", "churn", "Attrition", "Target", "ClosedAccount", "label"]
LOCAL_CLIENT_INDEX = "first"  'first' أو رقم صحيح
USE_SMOTE = False

 ------------------------- دوال مساعدة -----------------------------
def find_target_column(df, candidates=TARGET_CANDIDATES):
    for c in candidates:
        if c in df.columns:
            return c
    return None

def get_feature_names_from_column_transformer(ct, input_features):
    try:
        return list(ct.get_feature_names_out(input_features))
    except Exception:
        feature_names = []
        for name, transformer, cols in ct.transformers:
            if name == "num":
                feature_names.extend(cols)
            elif name == "cat":
                # حاول استخراج أسماء OneHot
                try:
                    onehot = transformer.named_steps.get("onehot")
                except Exception:
                    onehot = None
                if onehot is not None and hasattr(onehot, "get_feature_names_out"):
                    try:
                        feature_names.extend(list(onehot.get_feature_names_out(cols)))
                    except Exception:
                        feature_names.extend([f"cat_{c}" for c in cols])
                else:
                    feature_names.extend([f"cat_{c}" for c in cols])
        if not feature_names:
            # fallback عام
            transformed = ct.transform(pd.DataFrame([dict.fromkeys(input_features, 0)]))
            feature_names = [f"f_{i}" for i in range(transformed.shape[1])]
        return feature_names

# ------------------------- قراءة البيانات ---------------------------
if not os.path.exists(INPUT_FILE):
    print(f"خطأ: لم أجد الملف '{INPUT_FILE}' في المجلد الحالي ({os.getcwd()}).")
    sys.exit(1)

print("قراءة ملف الإكسل...")
df = pd.read_excel(INPUT_FILE, sheet_name=SHEET_NAME)
print(f"أعاد الملف {df.shape[0]} سطر و {df.shape[1]} عمود.")

target_col = find_target_column(df)
if target_col is None:
    print("لم أجد عمود هدف معروف تلقائياً. الرجاء التأكد من وجود عمود هدف مثل 'Churn'.")
    sys.exit(1)

print(f"عمود الهدف: {target_col}")

 محاولة تحويل القيم الثنائية النصية إلى 0/1
if df[target_col].dtype == object or df[target_col].dtype.name == "category":
    vals = df[target_col].dropna().unique()
    lower = set([str(v).lower() for v in vals])
    mapping = {}
    if lower <= {"yes", "no"}:
        mapping = {v: (1 if str(v).lower() == "yes" else 0) for v in vals}
    elif lower <= {"y", "n"}:
        mapping = {v: (1 if str(v).lower() == "y" else 0) for v in vals}
    elif lower <= {"true", "false"}:
        mapping = {v: (1 if str(v).lower() == "true" else 0) for v in vals}
    if mapping:
        df[target_col] = df[target_col].map(mapping)
    else:
        try:
            df[target_col] = df[target_col].astype(int)
        except Exception:
            print("تحذير: عمود الهدف غير ثنائي صريح؛ تابع بحذر.")

X = df.drop(columns=[target_col])
y = df[target_col]

numeric_cols = X.select_dtypes(include=["int64", "float64", "int32", "float32"]).columns.tolist()
categorical_cols = X.select_dtypes(include=["object", "category", "bool"]).columns.tolist()

print("الأعمدة الرقمية:", numeric_cols)
print("الأعمدة الفئوية:", categorical_cols)

------------------------- تقسيم ----------------------------------
if y.nunique() > 1:
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=TEST_SIZE, stratify=y, random_state=RANDOM_STATE)
else:
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=TEST_SIZE, random_state=RANDOM_STATE)

 ------------------------- preprocessing --------------------------
num_pipe = Pipeline([("imputer", SimpleImputer(strategy="median")), ("scaler", StandardScaler())])
 إذا كانت sklearn تعطي تحذير حول sparse، نحدد sparse=False
cat_pipe = Pipeline([("imputer", SimpleImputer(strategy="most_frequent")), ("onehot", OneHotEncoder(handle_unknown="ignore", sparse=False))])

preprocessor = ColumnTransformer([("num", num_pipe, numeric_cols), ("cat", cat_pipe, categorical_cols)])

print("تعلّم محول الميزات (fit preprocessor)...")
preprocessor.fit(X_train)
feature_names = get_feature_names_from_column_transformer(preprocessor, X_train.columns.tolist())
print(f"عدد الميزات بعد التحويل: {len(feature_names)}")

X_train_t = preprocessor.transform(X_train)
X_test_t = preprocessor.transform(X_test)
 تحويل إلى dense numpy إن لزم
if hasattr(X_train_t, "toarray"):
    X_train_t = X_train_t.toarray()
if hasattr(X_test_t, "toarray"):
    X_test_t = X_test_t.toarray()

 ------------------------- SMOTE اختياري --------------------------
if USE_SMOTE:
    if SMOTE is None:
        print("تحذير: imbalanced-learn غير مثبت؛ تجاهل SMOTE.")
    else:
        sm = SMOTE(random_state=RANDOM_STATE)
        X_train_t, y_train = sm.fit_resample(X_train_t, y_train)

 ------------------------- تدريب GridSearch -----------------------
print("تشغيل GridSearch على RandomForest...")
rf = RandomForestClassifier(random_state=RANDOM_STATE, class_weight="balanced")
param_grid = {"n_estimators": [100, 200], "max_depth": [6, 10, None], "min_samples_split": [2, 5]}
cv = StratifiedKFold(n_splits=3, shuffle=True, random_state=RANDOM_STATE)
grid = GridSearchCV(rf, param_grid, scoring="roc_auc", cv=cv, n_jobs=-1, verbose=0)
grid.fit(X_train_t, y_train)
best_rf = grid.best_estimator_
print("أفضل معاملات:", grid.best_params_)

 ------------------------- تقييم ---------------------------------
print("تقييم على مجموعة الاختبار...")
y_pred = best_rf.predict(X_test_t)
if hasattr(best_rf, "predict_proba"):
    try:
        y_proba = best_rf.predict_proba(X_test_t)[:, 1]
    except Exception:
        y_proba = y_pred
else:
    y_proba = y_pred

metrics = {}
try:
    metrics["AUC"] = roc_auc_score(y_test, y_proba)
except Exception:
    metrics["AUC"] = float("nan")
for name, fn in [("F1", f1_score), ("Recall", recall_score), ("Precision", precision_score), ("Accuracy", accuracy_score)]:
    try:
        metrics[name] = fn(y_test, y_pred)
    except Exception:
        metrics[name] = float("nan")

print("مقاييس الأداء:")
for k, v in metrics.items():
    print(f" - {k}: {v if not (isinstance(v, float) and np.isnan(v)) else 'N/A'}")

 ------------------------- SHAP -----------------------------------
print("حساب SHAP (TreeExplainer)...")
 تحويل إلى DataFrame مسماة إن أمكن
X_train_t_df = pd.DataFrame(X_train_t, columns=feature_names)
X_test_t_df = pd.DataFrame(X_test_t, columns=feature_names)

explainer = shap.TreeExplainer(best_rf, data=X_train_t_df)
shap_values = explainer.shap_values(X_test_t_df)

 اختيار قيم الفئة الإيجابية إذا كانت list
if isinstance(shap_values, (list, tuple)) and len(shap_values) >= 2:
    shap_vals_pos = shap_values[1]
    expected_value = explainer.expected_value[1] if isinstance(explainer.expected_value, (list, np.ndarray)) else explainer.expected_value
else:
    shap_vals_pos = shap_values if not isinstance(shap_values, (list, tuple)) else shap_values[0]
    expected_value = explainer.expected_value

 حساب متوسط |SHAP|
mean_abs_shap = np.abs(shap_vals_pos).mean(axis=0)
fi = pd.DataFrame({"feature": X_test_t_df.columns, "mean_abs_shap": mean_abs_shap}).sort_values("mean_abs_shap", ascending=False)
fi.to_csv("mean_abs_shap.csv", index=False)
print("تم حفظ mean_abs_shap.csv (أهمية الميزات العددية).")

 محاولة رسم وحفظ وفق توافر المكتبات
if HAS_MATPLOTLIB:
    try:
        import matplotlib.pyplot as plt
        import seaborn as sns  # قد لا يكون مثبتا؛ المخاطرة هنا أقل لأن HAS_SEABORN تفحص من قبل
        plt.figure(figsize=(10, 8))
        sns.barplot(data=fi.head(30), x="mean_abs_shap", y="feature")
        plt.title("Feature importance (mean |SHAP|) — top 30")
        plt.tight_layout()
        plt.savefig("shap_feature_importance_bar.png", dpi=300)
        plt.close()
        print("تم حفظ shap_feature_importance_bar.png (matplotlib).")
    except Exception as e:
        print("فشل رسم matplotlib:", e)
        HAS_MATPLOTLIB = False

elif HAS_PLOTLY:
    try:
        import plotly.express as px
        fig = px.bar(fi.head(30), x="mean_abs_shap", y="feature", orientation="h", title="Feature importance (mean |SHAP|) — top 30")
        fig.update_layout(yaxis={"categoryorder":"total ascending"})
        fig.write_html("shap_feature_importance_bar_plotly.html")
        print("تم حفظ shap_feature_importance_bar_plotly.html (plotly).")
    except Exception as e:
        print("فشل رسم plotly:", e)
        HAS_PLOTLY = False
else:
    print("لا توجد مكتبات رسم مثبتة؛ قمت بحفظ أهمية الميزات في CSV.")

 summary plot: حاول matplotlib أولاً ثم بديل حفظ بيانات للعرض
if HAS_MATPLOTLIB:
    try:
        plt.figure(figsize=(10, 6))
        shap.summary_plot(shap_vals_pos, X_test_t_df, show=False)
        plt.tight_layout()
        plt.savefig("shap_summary.png", dpi=300)
        plt.close()
        print("تم حفظ shap_summary.png (matplotlib).")
    except Exception as e:
        print("فشل رسم summary matplotlib:", e)

 حفظ بيانات SHAP الخام (مفيدة للعرض في أدوات أخرى)
np.save("shap_values_pos.npy", shap_vals_pos)
X_test_t_df.head(5).to_csv("X_test_t_sample.csv", index=False)
print("حفظت shap_values_pos.npy و X_test_t_sample.csv")

 تفسير محلي (force plot تفاعلي) — shap.force_plot مع matplotlib=False يعود كـ HTML (لا يحتاج matplotlib)
if LOCAL_CLIENT_INDEX == "first":
    local_idx = 0
else:
    try:
        local_idx = int(LOCAL_CLIENT_INDEX)
    except Exception:
        local_idx = 0

if local_idx < 0 or local_idx >= X_test_t_df.shape[0]:
    print("مؤشر العميل خارج النطاق؛ اخترت 0.")
    local_idx = 0

try:
    shap.initjs()
     expected_value قد يكون قائمة أو رقم
    ev_to_use = expected_value if not isinstance(expected_value, (list, np.ndarray)) else (expected_value[1] if len(expected_value) > 1 else expected_value[0])
    force_obj = shap.force_plot(ev_to_use, shap_vals_pos[local_idx, :], X_test_t_df.iloc[local_idx, :], matplotlib=False)
    html_fname = f"force_plot_client_{local_idx}.html"
    shap.save_html(html_fname, force_obj)
    print(f"تم حفظ تفسير محلي تفاعلي: {html_fname}")
except Exception as e:
    print("فشل إنشاء force_plot تفاعلي:", e)
     كبديل: حفظ الـ SHAP للقيم المحلية كـ JSON
    local_shap = dict(feature=list(X_test_t_df.columns), shap_values=shap_vals_pos[local_idx, :].tolist())
    with open(f"local_shap_client_{local_idx}.json", "w", encoding="utf-8") as fh:
        json.dump(local_shap, fh, ensure_ascii=False, indent=2)
    print(f"حفظت ملف JSON بديل: local_shap_client_{local_idx}.json")

print("\nانتهى التنفيذ. الملفات الناتجة في المجلد الحالي:")
for f in os.listdir("."):
    if f.startswith("shap_") or f.startswith("force_plot_client_") or f.endswith(".csv") or f.endswith(".npy") or f.endswith(".json"):
        print(" -", f)
